### DISCUSSION TITLE
Suggestions about collision response.

### DISCUSSION THREAD

**Discussion by yosoyfreeman** [godot-jolt/godot-jolt#812] [General]
Hi!

Regarding problematic collisions,  there are, in my eyes, two clearly identified issues:

- Move and slide will always limit the movement along the cross product of the last wall and the new collided wall. The problem here is that there is no verification that the new plane is totally or nearly identical to the last one, so this cross product can go to zero and stop all motion when the velocity is projected on it (The fact that it assumes the walls are infinite is another issue i would like to focus in another moment).

This points us toward another issue: Ghost collisions.  More often than not, we can collide with the same plane two times in a row even if it should not be possible. 

Move and slide it's is own thing, so i will focus on the Godot jolt side of things.

The minimum depth threshold has for sure showed useful to alleviate issues, but has introduced the problem of missing collisions.

What i propose is another way of contact validation for `_body_motion_collide`. Instead of discarding contacts based on their depth, we can get rid of multiple issues by discarding contacts that matches the following rules:

- If the collision normal is parallel (regardless of the orientation) to the motion vector we know its not a valid contact plane, as you could continue the motion parallel to the plane infinitely without ever stopping. This would cause the body to not get stuck because a collision parallel to the movement and reduce the number of times the same plane can be  processed two times in move and slide, triggering the stop_all_motion behaviour.

- If the collision plane points towards the motion vector it is not a valid plane, as you can not be stopped by something that is on your back while moving forward. This should allow the body to be released from the wall in the case you overlap slightly, because at the moment there is a velocity pointing outside the wall the wall would stop taking participation on the collision response.

Still with this changes move and slide needs improvements in which i'm trying to work on with mi limited resources, but they should improve the situation while making sure we don't discard any valid collision.

I was  going to propose this changes earlier but i wanted to be sure. All the research i did fight this type of problems in this two ways.  We can see something very similar here in the excellent Jolt CharacterVirtual:

https://github.com/jrouwe/JoltPhysics/blob/master/Jolt/Physics/Character/CharacterVirtual.cpp#L675-L681

```
// We don't want parallel or anti-parallel normals as that will cause our cross product below to become zero. Slack is approx 10 degrees.
float dot = other_normal.Dot(plane_normal);
if (dot < 0.984f && dot > -0.984f)
{
highest_penetration = penetration;
other_constraint = *c;
} 

```
https://github.com/jrouwe/JoltPhysics/blob/master/Jolt/Physics/Character/CharacterVirtual.cpp#L754-L771

```
// Flag contacts as having a collision if they're close enough but ignore contacts we're moving away from.
// Note that if we did MoveShape before we want to preserve any contacts that it marked as colliding
for (Contact &c : mActiveContacts)
if (!c.mWasDiscarded
	&& !c.mHadCollision
	&& c.mDistance < mCollisionTolerance
	&& (inSkipContactVelocityCheck || c.mSurfaceNormal.Dot(mLinearVelocity - c.mLinearVelocity) <= 1.0e-4f))
{
	if (ValidateContact(c) && !c.mIsSensorB)
		c.mHadCollision = true;
	else
		c.mWasDiscarded = true;
}

```
https://github.com/jrouwe/JoltPhysics/blob/master/Jolt/Physics/Character/CharacterVirtual.cpp#L177

`&& inResult.mPenetrationAxis.Dot(mDisplacement) > 0.0f) // Ignore penetrations that we're moving away from`

This would have the side effect of having to move towards a wall to get the information about it. But that is because move and slide mixes the collisions to which it should react with the collisions used to gather information about the environment, but that should be improved by it's own. 

PseudoCode:
```

if hit.normal.dot(motion.normalized) >= 0 : //Discard planes on your back.
    continue

if is_zero_approx(motion.normalized.dot(hit.normal)): //Discard parallel planes.
    continue

```

PS: I seen that is a recurring pattern using the sweep test to collect information about static collisions in Godot, so may i ask, is there a specific method other than _body_motion_collide to gather collision data? Because it is not possible to satisfy the necessities of a movement check and a collision gathering check with the same rules, and may be one of the issues forcing this constant balancing between correct sliding and correct data gathering.

**Comment by mihe** [comment]
Thank you for the in-depth suggestions.

Again, I haven't really spent enough time reading up on or implementing character movers that I can speak to the validity of your suggestions, nor the parallels you draw to Jolt's `CharacterVirtual`, but on the face of it they make sense to me, and from some preliminary/basic testing they do seem to be a decent replacement for the current contact depth threshold hack.

You'll find the implementation in [this](https://github.com/godot-jolt/godot-jolt/compare/master...mihe:godot-jolt:new-contact-discard) branch.

It performs ever so slightly worse in the test scene provided in #529, but frankly it's close enough that it's hard to tell if it's just different rather than worse. When ripping out `stop_all_motion` from the Godot side of things it's close to flawless in terms of not getting stuck, much like the current hack, so I'm assuming it's just about improving things there.

It does fix the performance regression (#626) introduced by the current contact depth threshold, since there's no longer any need for `move_and_slide` to perform its `apply_floor_snap` due to being pushed too far away from the floor every single time. I believe this has caused other (unreported) ill effects as well, so this change alone might be reason enough to go with your suggestions.

I am however not seeing a huge difference in its ability to detect walls (#758), although there might be some improvement. I'm assuming this is what you're referring to when you say "having to move towards a wall to get the information about it". This phenomenon is present in Godot Physics as well to some degree, but is (from what I understand) supposed to be alleviated (in a somewhat odd way) by leaving the depenetration ever so slightly short.

I am also seeing a new discrepancy when comparing with Godot Physics, which is that the number of slide iterations actually go down from 2 to 1 when pushing up against a wall. I'll need to look into where that comes from to see whether it's a good or bad thing.

> I seen that is a recurring pattern using the sweep test to collect information about static collisions in Godot, so may i ask, is there a specific method other than _body_motion_collide to gather collision data?

I'm not sure what you mean by sweep test in this context. There is no traditional sweep test in Godot, nor in Godot Jolt. Any shape-casting is done using a binary search, by stretching the shape by the cast distance along its support vector a number of times until you've found a spot where you didn't collide and a spot where you did collide, which are the "safe" and "unsafe" fractions. This binary search is (unfortunately) more or less required by any physics implementation in Godot due to needing to output these two fractions.

The only part of `body_test_motion` in Godot Jolt that uses this shape-casting is `_body_motion_cast`. The other two steps, meaning `_body_motion_recover` and `_body_motion_collide`, just use a regular collision query, much like Godot Physics.

**Reply by yosoyfreeman** [reply]
Thank you a lot for looking into it. The new branch seems good to me and I'm happy it seems to be an improvement. Have you tried to discard parallel and anti parallels too? They are illegal for the same reason. If you try to parametrize the time of impact with a plane when the movement is parallel to the plane it would be impossible, because they would never collide. 

> Again, I haven't really spent enough time reading up on or implementing character movers that I can speak to the validity of your suggestions, nor the parallels you draw to Jolt's CharacterVirtual

I try my best to research a topic when i'm trying to collaborate to it, but you are absolutely right in taking my suggestion with a grain of salt, as sadly my socio-economic circustances have prevented me from acquiring any formal education on the matter. However, while mathematics is not my strong point i been familiarizing myself with this kind of algorithms for about 4 years now, since i started reporting issues with Godot physics in godot 3.

In reference to the parallels with CharacterVirtual, It is in all practical effects a kinematic body with an sliding algorithm and its behaviour is directly comparable with our CharacterBody3D. The difference, of course, is on the complexity, precision and extra features, such as simulating contacts with rigidbodies and react to them, which is what may make it look like a fully physics based controller while is not. 

Another fundamental difference that aims to solve the same problem i was working on is the way in which it handles multiple planes for the stuck cases inherent to sliding algorithms such as the corner cases.  In this case, character virtual is taking contact points and converting them into constrains. Extra constrains are created too to provide different behaviours such as cancelling the velocity towards steep slopes (The equivalent to our stop on wall). The most interesting part is that, instead of reacting to a single plane at a time and reacting to it (Which we could call the first version of this kind of algorithms) or always limiting the axis of freedom based on previous collisions, it will resolve the constrains to get a new velocity. This new velocity is tested against the normals (constraints) of the previous iteration (Edit: I can't remember for sure if its the previous iteration. I believe is done in the same iteration with the same contacts. We resolve the nearest one and then see if we are violating another one as a consequence of that), in the case the new velocity is violating one of the previous contacts (This is for example the infamous case of sliding on a ramp, touching a ceiling, going to the ramp etc and never resolving) it will check for the contact that we are violating the most (We are moving against the previous normal, so we are stuck ). If this is the case, it will simply take the cross product of the first constrain and the second one we selected to create the vector in which project the velocity. What i mean is that CharacterVirtual is probably the most complete source of information about kinematic controllers available out there and we can learn from it as its contents can be directly translated to our use cases.

> When ripping out stop_all_motion from the Godot side of things it's close to flawless in terms of not getting stuck, much like the current hack, so I'm assuming it's just about improving things there.

Yes, as i mentioned at some point before, i find the logic of that piece of code to be just wrong. Yes, it can help resolving cases of acute corners, but it will drastically affect the outcome if anything goes slightly wrong. I can make a detailed description of what move and slide is exactly doing and what points i found to be the most relevant to tackle if that is of some help.

> I am however not seeing a huge difference in its ability to detect walls (https://github.com/godot-jolt/godot-jolt/issues/758), although there might be some improvement. I'm assuming this is what you're referring to when you say "having to move towards a wall to get the information about it". This phenomenon is present in Godot Physics as well to some degree, but is (from what I understand) supposed to be alleviated (in a somewhat odd way) by leaving the depenetration ever so slightly short.

Exactly, because we are now ignoring planes that we could not collide with along the motion (And that's the right thing to do on a motion test), contacts with the walls will no longer be reported if you are not moving against them. The problem here is that relying purely of the collisions we use to handle movement to gather the data is extremely limiting, as we are forced to find a balance between breaking both things even if its working exactly as expected. Either we use the correct contacts to alter the motion, missing information about the environment, or we gain information about the environment by breaking the collision on purpose (Not fully un-stucking the body or accepting illegal collisions on the motion test, which often leads to a full stuck). This is related to my question about using _body_motion_collide for static checks.

I'm sorry if i confused the terminology, i understand that for "traditional" sweep test you are referring to the classic way of testing a shape against all near polygons directly.  I'm aware we don't have those, but the part of a move and slide algorithm that moves the body to the nearest point, with whatever method, is often referred to as the sweep test pass. This is referred to in this way in Jolt too, specifically here: 
`
bool CharacterVirtual::GetFirstContactForSweep(RVec3Arg inPosition, Vec3Arg inDisplacement, Contact &outContact, const IgnoredContactList &inIgnoredContacts, const BroadPhaseLayerFilter &inBroadPhaseLayerFilter, const ObjectLayerFilter &inObjectLayerFilter, const BodyFilter &inBodyFilter, const ShapeFilter &inShapeFilter) const`

Made the clarification, my point is that is impossible to get reliable collision information about the environment (The wall without touching pushing it, for example) with a reliable motion test, as their purposes are vastly different. A motion test should only care about the collisions it should be stopped by, which leaves out any relevant information about the things you are _touching_ but do not prevent you for moving further.

This is a limitation in move and slide. And because move and slide is done this way, all the physics API is adapted to accommodate it. What needs to be done in move and slide is a better implementation overall. This limitation would not be there if Godot had two proper differentiate test (Which jolt has). This is: A test for checking collisions along the path/sweep test, as showed before, and another test to gather all the collision data itself in a static test that is not aware of motion. An example in Jolt of this two would be:

For the motion test:

`bool CharacterVirtual::GetFirstContactForSweep(RVec3Arg inPosition, Vec3Arg inDisplacement, Contact &outContact, const IgnoredContactList &inIgnoredContacts, const BroadPhaseLayerFilter &inBroadPhaseLayerFilter, const ObjectLayerFilter &inObjectLayerFilter, const BodyFilter &inBodyFilter, const ShapeFilter &inShapeFilter) const  `


For the collision test:

`void CharacterVirtual::CheckCollision(RVec3Arg inPosition, QuatArg inRotation, Vec3Arg inMovementDirection, float inMaxSeparationDistance, const Shape *inShape, RVec3Arg inBaseOffset, CollideShapeCollector &ioCollector, const BroadPhaseLayerFilter &inBroadPhaseLayerFilter, const ObjectLayerFilter &inObjectLayerFilter, const BodyFilter &inBodyFilter, const ShapeFilter &inShapeFilter) const`

or 
`
void CharacterVirtual::GetContactsAtPosition(RVec3Arg inPosition, Vec3Arg inMovementDirection, const Shape *inShape, TempContactList &outContacts, const BroadPhaseLayerFilter &inBroadPhaseLayerFilter, const ObjectLayerFilter &inObjectLayerFilter, const BodyFilter &inBodyFilter, const ShapeFilter &inShapeFilter) const`

i must recognize that i struggle to fully understand the difference between the last two. What i'm trying to say is that is way, way, way easier to just improve or replace move and slide that it is to battle with a balance of two things that have two different purposes. 

Annoyingly, get_rest_info, the method you would use for getting this information will only provide one contact point and lacks a margin, but using a ShapeCast node with a motion of zero gives this information.  This is counter intuitive and a limitation, at least for me, that i prototype in GDScript, as i'm good with systems, not so much with programming technicalities. I made a proposal to improve that in here: https://github.com/godotengine/godot-proposals/issues/9341but i have to say that i doubt to which extends maintaining the current api, which was made by people who no longer collaborate nor maintain it is a better idea than adapting the Jolt equivalents, which are open source, maintained and well documented. But i guess that's for another discussion. 

Long history short: We can fix move and slide, but for that i need the two proper test: One for the sweep that ignores all the collisions that would be impossible to stop it, and another one which has no motion parameter but provides you all the contacts, similar to get rest info. In that way we can do a better collision resolution while being able to gather all the relevant information about the walls, the floor and the ceiling. This is how Jolt does it, and with some luck we can make some improvements in a problem that has been present for years now.

Thank you for your time;

Freeman.

**Reply by mihe** [reply]
> Have you tried to discard parallel and anti parallels too?

The code in that branch (`motion.normalized().dot(normal) >= -CMP_EPSILON`) should cover both of your suggested if-statements, unless I'm mistaken.

> I try my best to research a topic when i'm trying to collaborate to it, but you are absolutely right in taking my suggestion with a grain of salt, as sadly my socio-economic circustances have prevented me from acquiring any formal education on the matter. However, while mathematics is not my strong point i been familiarizing myself with this kind of algorithms for about 4 years now, since i started reporting issues with Godot physics in godot 3.

Fair enough! My point was less about your capabilities and more about my own, or lack thereof, in case you relied on me to confirm any of this.

> Yes, as i mentioned at some point before, i find the logic of that piece of code to be just wrong. Yes, it can help resolving cases of acute corners, but it will drastically affect the outcome if anything goes slightly wrong. I can make a detailed description of what move and slide is exactly doing and what points i found to be the most relevant to tackle if that is of some help.

I've been tempted to submit a PR to Godot to simply remove the `stop_all_motion` portion of `move_and_slide`. I'm curious how you would feel about this? It seems to me that it causes more problems than it solves at the moment, so removing it seems like a good idea to me, but I struggle to see the full impact of that change.

> the part of a move and slide algorithm that moves the body to the nearest point, with whatever method, is often referred to as the sweep test pass

Got it. I misunderstood then.

> A test for checking collisions along the path/sweep test, as showed before, and another test to gather all the collision data itself in a static test that is not aware of motion.

Are you talking about tests within `move_and_slide` itself, or on the physics server side of things? Because this split is how things work in `PhysicsServer3D.body_test_motion` already.

In Godot Jolt (which is largely just a cleaned up port of the Godot Physics implementation) you first have the `_body_motion_recover` method which does the iterative depenetration, using a normal collision check without any motion incorporated, the same collision check as what's done in Jolt's `CharacterVirtual::CheckCollision`.

Then it does the actual sweep, meaning `_body_motion_cast`, which gives you the unobstructed distance, which can sort of be mapped to Jolt's `CharacterVirtual::GetFirstContactForSweep`, although Jolt uses a "proper" sweep test here, which yields contacts, whereas in Godot you have to perform a separate collision check at the resulting fraction instead.

This last separate collision check would be `_body_motion_collide`, which again is just a normal collision query, the same as what's used in `_body_motion_recover`, and which again doesn't incorporate any motion. It is a static test, as you put it, and not much different from `get_rest_info`.

Without looking at the Jolt code too closely I feel like the high-level components are all somewhat similar, but done in a different order, and I'm sure done in a much better way in Jolt.

> i have to say that i doubt to which extends maintaining the current api, which was made by people who no longer collaborate nor maintain it is a better idea than adapting the Jolt equivalents, which are open source, maintained and well documented. But i guess that's for another discussion.

If/when we get to the point where Godot Physics 3D is deprecated and eventually replaced by Jolt there might be some discussion to be had about this, but I must admit I would find it somewhat sad considering how the API is currently fairly "open" and can accommodate different physics engines without too much trouble.

**Reply by yosoyfreeman** [reply]
Hi! Thank you for your time.

> The code in that branch (motion.normalized().dot(normal) >= -CMP_EPSILON) should cover both of your suggested if-statements, unless I'm mistaken.

Not completely. It covers the case in which a plane points _towards_  the motion(The dot product of the normal and the normalized motion is  a bit more than 0). However, it will allow planes that are _parallel_ to the motion, because zero is not included, and if we take into consideration numerical error, the dot product of the parallel plane could be ever so slightly negative. It could be done tweaking a single dot, but i have the feeling that tweaking it will be simpler be keeping this two check separated.

![imagen](https://github.com/godot-jolt/godot-jolt/assets/64601424/6747516d-2021-429e-a4ec-bd3d87f3bfb1)

If we describe the rule to ignore the pink plane as "Ignore planes that you are leaving behind", the green and blue rules would be "Ignore planes that you are _touching_, but can't stop you"

A similar technique is used in [this paper](https://www.peroxide.dk/papers/collision/collision.pdf) for the sweep test, to remove a good amount of illegal contacts. It is present (For similar reason, but not exactly in the same case) in the example i shown of jolt code.

For the threshold, i recommend something between 5 and 10 degrees of margin to mark contacts as invalid. In pseudocode:

```
if motion.normalized.dot(hit.normal) < (-1 + half the margin) or motion.normalized.dot(hit.normal) > (1 - half the margin):
    continue
```

This should, theoretically, improve the cases in which move and collide stops while moving parallel to a wall. Most of this full stops are caused because the minimal penetration, even due to numerical error, can detect the wall you are moving parallel to and set the safe fraction to 0 or almost zero. This improvement should be reflected in move and slide too.

> I've been tempted to submit a PR to Godot to simply remove the stop_all_motion portion of move_and_slide. I'm curious how you would feel about this? It seems to me that it causes more problems than it solves at the moment, so removing it seems like a good idea to me, but I struggle to see the full impact of that change.

Even thou that piece of code is executed in a non so elegant way, for saying it politely, it is there to prevent an inherit problem with sliding algorithms, which is referred to the acute corner case. You see, most of the times, the most simple implementation of an sliding algorithm will work flawlessly, but there is one exception to this rule. The basic algorithm goes like this:

- calculate a delta velocity. This is the motion for the frame.

- Move the body towards the collision point, minus a margin (Actually touching the geometry is considered a bad practice because it WILL lead to bugs and more expensive calculations, as a result of constantly colliding with the geometry on the first iteration and having to do an extra projection and cast)

- Remove the travel from the delta velocity (get_remainder)

- Project the remaining velocity into the sliding plane (This is the collision normal, **not** the surface normal)

- Repeat with the new velocity and direction until we don't have remaining motion.

The algorithm relies on the natural reduction of motion as a result of the projection to accomplish it's task. This works great most of the times, and that's why you will almost always see collisions resolved in 2 to 4 iterations with this method.

So, why is there a limit of iterations? As we said, we rely on the reduction of motion as a result of the projection. This means that, the more parallel the planes are (The more acute the corner) less the reduction of delta velocity per iteration. If the corner is acute enough and the motion big enough, this will result in so much iterations it can hard freeze your PC.  Sadly, the most known and used solution for that is in [this paper](https://arxiv.org/pdf/1211.0059.pdf) which brute force a solution by assuming every single plane you touched during the motion is infinite... Which result in the body no longer sliding properly along surfaces. From there it come the infamous `block all motion` which adds the extra problem of just doing it with walls, causing the algorithm to both not slide properly and still get jitter. Yes, it "fixes" the case of getting stuck because the acute corner... It breaks everything else. 

In the prototype of my implementation i solve this issue by actually reacting to multiple collisions per iteration. That is, if i collide with two planes **at the same time** I'll only allow movement between the cross product of the two (And of course, making sure the two collisions actually imply a real conflict on the movement). Jolt fixes this by trying to solve with the nearest collision, and then seeing if the result of solving against that violates any other contact point. If it does, it picks the one it violates the most and perform the cross product of the two.

Here is my prototype, and as you can see the body slides smoothly without stutter or getting stuck. The end of the video shows my implementation failing because the minimum threshold that is currently on the release fails to detect the ceiling an the floor at the same time, which is a good show if what happens when you don't solve the corner issue correrctly (In this case is contemplated, but fails to recognize the ceiling). This is just a prototype, i wanted to see if i could offer some help with collisions before continuing and being able to do a proposal for a replacement of move and slide.

https://github.com/godot-jolt/godot-jolt/assets/64601424/3bb97169-7d55-426f-a9db-3a9ef3d1093d

So ad bad as the actual fix is, it is what is preventing the acute corner scenario, which would imply another kind of stuck. Fixing it is totally doable (But i must say i been unsuccessful making the team see the deep problems of the actual approach), but for that we first need to clean the problems with the collision issues and make sure we have all of this working right. I know for sure the person who made move and slide worked hard, but time was scarce, the engine very new and lots of physics issues were not identified or fixed, so they made the best to get a good result with a physics engine that was failing them. In some cases this went to the point of using this fails as features, for example the way walls are check by not correctly de penetrating the body.  Now that physics are better and improvements are being done, the parts of the code that relied on the issues to work are starting to ironically fail, see for example how as we improve movement, we lost precision in wall and ceiling detection while standing still.

> Are you talking about tests within move_and_slide itself, or on the physics server side of things? Because this split is how things work in PhysicsServer3D.body_test_motion already.

Both. Move and slide is gathering information only by the "movement test". More on the second later.

Thank you **a lot** for the detailed explanation. It really helps me understand the full process better. I'll try to clarify what i meant better.

Right now, motion methods perform as you said two things as the last two steps: It will find a safe fraction (Which we are trying to make the longer valid one by ignoring contacts that should not stop us, like the checks we talked), then performs another check to gather the information at the point of contact. You can modify what contacts to ignore in which phase, but the user cant, which is the reason people use a _motion check_ with zero movement to check for surrounding contacts. You are doing an excellent job making sure we ignore the correct contacts for the "sweep" test, but now we are gathering **mixed**  information. We are getting the contacts that made the body stop, which is useful information by itself AND all the contacts on the check done on the safe fraction point, without a way to discern which are which. This difference is exactly the difference between the planes we want to react to, and the planes we want to tell the user they are touching. That's why i meant by motion or static checks. When i do a motion check i expect to get the planes that stopped the motion, and only those, for example, a wall. When i do a general/static check i expect to get the floor and the ceiling too, as you said, similar to get_rest_info, but get rest info ONLY gives one contact.

In Godot that information is mixed, which is not bad by itself, but then the name test_motion no longer fully match the expected behavior. That's why i think that motion methods should only return the planes that stopped you, as they do in other engines, to get the information we need from them, and we should be able to use get_rest_info to get all the contact points.

For not breaking compatibility and keep things simple, i been thinking that, what we could do for the moment is to check in PhysicsServer3D.body_test_motion if there is motion at all. If there is motion, the rules are the one we said: You will get the contacts that are able to **stop you**, ignoring planes back to you or almost parallel to the motion. If there is no motion, you get all the contact points.

That way we can keep the current api unchanged, and still provide better results for everyone, because nobody expects to get the wall behind them or the floor above if they are moving forward, but the people who is already using a test move or move and collide with zero motion are for sure trying to get all the contact points. Is a change that will help everyone, and if documented will provide more flexibility to the users because they will know exactly what information they get, which is not clear on the documentation at the moment. And most importantly, it should not break any gameplay.

That way i could finally react correctly to the planes found on the movement, and calculate the new velocity with them, _and then_  do a collision check for all the contact points that i can classify between floors, walls and ceilings.

> If/when we get to the point where Godot Physics 3D is deprecated and eventually replaced by Jolt there might be some discussion to be had about this, but I must admit I would find it somewhat sad considering how the API is currently fairly "open" and can accommodate different physics engines without too much trouble.

My bad again. I just mean improving the methods exposed to the user and or creating new ones if needed, as we are doing right now making sure illegal planes are not reported or stop us.

Freeman.

**Reply by mihe** [reply]
> However, it will allow planes that are _parallel_ to the motion, because zero is not included, and if we take into consideration numerical error, the dot product of the parallel plane could be ever so slightly negative.

The code in my previous comment (what's in the branch) does include zero, and does include a slight numerical error, same as what's used in `is_zero_approx`. So `motion.normalized().dot(normal) >= -CMP_EPSILON` (note the negation) should catch the red, green and blue cases in your diagram, meaning parallel/anti-parallel planes (i.e. a dot product of roughly 0, with some tolerance) as well as perpendicular planes facing the same direction (i.e. a dot product of 1), and anything between those two of course.

> For the threshold, i recommend something between 5 and 10 degrees of margin to mark contacts as invalid.

I would prefer a small threshold to begin with, like `CMP_EPSILON`, meaning ~0.25 degrees. 5-10 degrees strikes me as a fairly high threshold.

> Even thou that piece of code is executed in a non so elegant way, for saying it politely, it is there to prevent an inherit problem with sliding algorithms, which is referred to the acute corner case. You see, most of the times, the most simple implementation of an sliding algorithm will work flawlessly, but there is one exception to this rule. The basic algorithm goes like this:

I appreciate the in-depth explanation. I'll leave it alone for now then.

> For not breaking compatibility and keep things simple, i been thinking that, what we could do for the moment is to check in PhysicsServer3D.body_test_motion if there is motion at all. If there is motion, the rules are the one we said: You will get the contacts that are able to **stop you**, ignoring planes back to you or almost parallel to the motion. If there is no motion, you get all the contact points.

Right, that should be how things work in my branch already. It doesn't make a whole lot of sense to be discarding anything based on the dot product of a zero vector either way.

> My bad again. I just mean improving the methods exposed to the user and or creating new ones if needed, as we are doing right now making sure illegal planes are not reported or stop us.

Ah, I misunderstood. I thought you meant switching over to `CharacterVirtual` or something. Nevermind. :)

Again, I appreciate the in-depth explanations to everything. I feel bad leaving a relatively short comment to such a detailed reply, but I'm not quite as deep into this as you clearly seem to be.

I'll investigate that decrease in slide iterations that I mentioned in my original reply, and assuming that isn't causing any odd discrepancies I'll go ahead and merge the branch.

**Reply by yosoyfreeman** [reply]
Hi Mihe!

> The code in my previous comment (what's in the branch) does include zero, and does include a slight numerical error, same as what's used in is_zero_approx. So motion.normalized().dot(normal) >= -CMP_EPSILON (note the negation) should catch the red, green and blue cases in your diagram, meaning parallel/anti-parallel planes (i.e. a dot product of roughly 0, with some tolerance) as well as perpendicular planes facing the same direction (i.e. a dot product of 1), and anything between those two of course.

Awesome! I apologize, didn't see the minus! And i only recently started learning source code, so the language is still a bit unfamiliar. I appreciate your patience in that regard! 

> I would prefer a small threshold to begin with, like CMP_EPSILON, meaning ~0.25 degrees. 5-10 degrees strikes me as a fairly high threshold.

You are absolutely right! 

> Right, that should be how things work in my branch already. It doesn't make a whole lot of sense to be discarding anything based on the dot product of a zero vector either way.

That's awesome! I didn't knew that, it is documented? The only reference on the docs to performing statics checks is using the ShapeCast node, so knowing that motion test will behave accordingly if there is no motion to them would be great.

> Ah, I misunderstood. I thought you meant switching over to CharacterVirtual or something. Nevermind. :)

Oh, no, my plan is to work on a better move and slide, and for that some of the techniques and patterns used in CharacterVirtual are going to be really useful, to have a production ready thing to compare against and learn from, but it should be done using Godot Api, so it works with any physics engine.

> Again, I appreciate the in-depth explanations to everything. I feel bad leaving a relatively short comment to such a detailed reply, but I'm not quite as deep into this as you clearly seem to be.

Don't feel bad! Your help is extremely useful and your explanations will make my life a lot easier, as i understand better how the tools I'm working with work. I was trying to gather all this information by myself and i would have taken my much, much longer.  I hope the new fix helps with the missing collisions and i can continue with the developing of the sliding algorithm! I really appreciate your work and that you took the time to consider my suggestions. I hope my verbosity didn't bothered you, I'm trying to improve on that :)

> I'll investigate that decrease in slide iterations that I mentioned in my original reply, and assuming that isn't causing any odd discrepancies I'll go ahead and merge the branch.

I can not be sure without seeing it, but if you are not applying gravity while moving against the wall, it is probable that the decrease in iterations is due to the floor being correctly ignored even if depenetrating the wall made you touch it or get inside it a little bit. If the body moves against a straight wall in a straight line, in theory all velocity should stop because the projection on the plane is zero, and that should take 1 iteration.

**Reply by mihe** [reply]
> That's awesome! I didn't knew that, it is documented? The only reference on the docs to performing statics checks is using the ShapeCast node, so knowing that motion test will behave accordingly if there is no motion to them would be great.

To clarify, I was referring to [this](https://github.com/godot-jolt/godot-jolt/compare/master...mihe:godot-jolt:new-contact-discard) yet-to-be-merged branch. As you can see there the discarding is only done if `p_motion.length_squared() > 0`, so passing in a motion of zero means nothing will be discarded.

I'm not sure what the `ShapeCast3D` node would provide you that the regular physics queries in `PhysicsDirectSpaceState3D` doesn't already. All that `ShapeCast3D` does is use the same `cast_motion` and `get_rest_info` methods that's available in `PhysicsDirectSpaceState3D`.

> I hope my verbosity didn't bothered you

Not at all!

> I can not be sure without seeing it, but if you are not applying gravity while moving against the wall, it is probable that the decrease in iterations is due to the floor being correctly ignored even if depenetrating the wall made you touch it or get inside it a little bit. If the body moves against a straight wall in a straight line, in theory all velocity should stop because the projection on the plane is zero, and that should take 1 iteration.

I haven't had time to look into this, but yes, I suspect you might be right about the depenetration causing some collision with the floor.

**Reply by yosoyfreeman** [reply]
> I'm not sure what the ShapeCast3D node would provide you that the regular physics queries in PhysicsDirectSpaceState3D doesn't already. All that ShapeCast3D does is use the same cast_motion and get_rest_info methods that's available in PhysicsDirectSpaceState3D.

I believe there is a difference between what can you do with get_rest_info via gdscript and natively. In GDScript, get_rest_info provides only one collision point and normal and no more. This is also indicated on the documentation, "If there is more than one contact, the closest one will be selected". So i can not use get_rest_info to get all the contact points i have to resolve against in GDScript. 

> Checks the intersections of a shape, given through a [PhysicsShapeQueryParameters3D](https://docs.godotengine.org/en/stable/classes/class_physicsshapequeryparameters3d.html#class-physicsshapequeryparameters3d) object, against the space. If it collides with more than one shape, the nearest one is selected. The returned object is a dictionary containing the following fields:

If we take a look at PhysicsShapeQueryParameters3D, we can see that no max collisions parameter is accepted: 
![imagen](https://github.com/godot-jolt/godot-jolt/assets/64601424/461ee7d6-9ae4-488d-a2cb-f3e51d497b0c)

So, if i want to get all the contact points, i have to use either move_and_collide with zero motion or a shapecast with zero motion. Hope the problem is clear now.

> I haven't had time to look into this, but yes, I suspect you might be right about the depenetration causing some collision with the floor.

Again, thank you so much! I'll be peeking here to see any updates in that regard. 

Freeman.

**Reply by mihe** [reply]
> I believe there is a difference between what can you do with get_rest_info via gdscript and natively. In GDScript, get_rest_info provides only one collision point and normal and no more. This is also indicated on the documentation, "If there is more than one contact, the closest one will be selected". So i can not use get_rest_info to get all the contact points i have to resolve against in GDScript.

There is no difference, as far as I'm aware. The way `ShapeCast3D` uses `get_rest_info` is by running it multiple times, and each time adding the `RID` of the body it hit to the `exclude` list of `PhysicsShapeQueryParameters3D`, until `get_rest_info` reports no more collisions, meaning an empty dictionary in the case of GDScript. So you might get multiple contact points from `ShapeCast3D`, but it'll only ever be one contact point per body.

The code itself is a pretty straight-forward read: https://github.com/godotengine/godot/blob/99ff024f78f65ba0bc54fb409cfeca43ba2008fe/scene/3d/physics/shape_cast_3d.cpp#L424-L433

It's a bit of an odd design, I must say.

**Reply by yosoyfreeman** [reply]
> There is no difference, as far as I'm aware. The way ShapeCast3D uses get_rest_info is by running it multiple times, and each time adding the RID of the body it hit to the exclude list of PhysicsShapeQueryParameters3D, until get_rest_info reports no more collisions, meaning an empty dictionary in the case of GDScript. So you might get multiple contact points from ShapeCast3D, but it'll only ever be one contact point per body.

That is unfortunate and explains why i always got bad results using the ShapeCast node. You know if there is any reason for using rest info for ShapeCast instead of doing it similarly to body motion collide and being able to get all the contact points?

**Reply by mihe** [reply]
> You know if there is any reason for using rest info for ShapeCast instead of doing it similarly to body motion collide and being able to get all the contact points?

Even if it's a scene node exposed by Godot itself it's still a scene node, so it's limited by the generic interface of `PhysicsDirectSpaceState3D`.

The better question is why the collision checks in `PhysicsDirectSpaceState3D` are designed/separated in the way they are, with `get_rest_info`, `collide_shape` and `intersect_shape` all providing different and incomplete parts of the collision check. I'm assuming it's for performance and/or legacy reasons.
